### PR TITLE
jwt: add middleware and token creation

### PR DESCRIPTION
**Pull request by he2lec**

### CODE REVIEW

File: src/controller/UserController.ts [hunk 1]
```diff
@@ -14,6 +16,7 @@ export class UserController {
 
   @Get('/users/:id')
   async one(@Param('id') id: number) {
+    console.log('Getting user id ' + id);
```
Comment (MrColorado) [inline src/controller/UserController.ts]:
```suggestion
```

**Comment by MrColorado** [comment]
The facebook authentification is still missing I think it is better to wait a MR for it before merging this one

File: src/config.ts [hunk 2]
```diff
@@ -1,7 +1,7 @@
 // Load config from `.env` file
 require('dotenv').config();
 
-function getEnv(variable: string, defaultValue = undefined): string {
+function getEnv(variable: string, defaultValue = <string | undefined> ''): string {
```
Comment (hugodelahousse) [inline src/config.ts]: You're casting an empty string into a <string | undefined>. What you probably want is:
```suggestion
function getEnv(variable: string, defaultValue: string? = ''): string {
```

File: src/middlewares/checkJwt.ts [hunk 3]
```diff
@@ -0,0 +1,25 @@
+import { Request, Response, NextFunction } from 'express'
+import * as jwt from 'jsonwebtoken'
+import config from 'config'
+
+export function checkJwt(request: Request, response: Response, next: NextFunction) {
+    const token = <string>request.headers.authorization;
```
Comment (hugodelahousse) [inline src/middlewares/checkJwt.ts]:
```suggestion
    const token: string = request.headers.authorization;
```

File: src/middlewares/checkJwt.ts [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+    let jwtPayload;
```
Comment (hugodelahousse) [inline src/middlewares/checkJwt.ts]: You should declare an interface with the contents of the jwtPayload to take advantage of TypeScript

File: src/middlewares/checkJwt.ts [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+    
+    //Verification of token
+    try {
+        jwtPayload = <any>jwt.verify(token, config.JWT_SECRET);
```
Comment (hugodelahousse) [inline src/middlewares/checkJwt.ts]:
```suggestion
        jwtPayload = jwt.verify(token, config.JWT_SECRET);
```

File: src/middlewares/checkJwt.ts [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+    }
+  
+    //New token of 1h on every requests
+    const { userId, googleId } = jwtPayload;
+    const newToken = jwt.sign({ userId, googleId }, config.JWT_SECRET, {
+        expiresIn: '1h'
+    });
+    response.setHeader('token', newToken);
+    next();
+  };
```
Comment (hugodelahousse) [inline src/middlewares/checkJwt.ts]:
```suggestion
  };
  
```

File: src/config.ts [hunk 7]
```diff
@@ -13,11 +13,13 @@ function getEnv(variable: string, defaultValue = undefined): string {
 interface Config {
   GOOGLE_CONSUMER_KEY: string;
   GOOGLE_CONSUMER_SECRET: string;
+  JWT_SECRET: string;
 }
 
 const config: Config = {
-  GOOGLE_CONSUMER_KEY: getEnv('GOOGLE_CONSUMER_KEY'),
-  GOOGLE_CONSUMER_SECRET: getEnv('GOOGLE_CONSUMER_SECRET'),
+  GOOGLE_CONSUMER_KEY: getEnv('GOOGLE_CONSUMER_KEY', 'key'),
```
Comment (hugodelahousse) [inline src/config.ts]: Why are you using these defaults values ?

File: package.json [hunk 8]
```diff
@@ -1,66 +1,69 @@
...
-  "scripts": {
-    "start": "ts-node -r tsconfig-paths/register src/index.ts",
-    "watch": "nodemon -e ts -w ./src -x npm run start",
-    "lint": "tslint -p tsconfig.json src/**.ts",
-    "fixtures": "ts-node -r tsconfig-paths/register node_modules/.bin/fixtures",
-    "migrations:run": "ts-node -r tsconfig-paths/register ./node_modules/.bin/typeorm migration:run",
-    "build": "tsc",
-    "test": "NODE_ENV=test mocha --timeout 5000 -r ts-node/register -r tsconfig-paths/register src/tests/**/*.test.ts"
-  }
-}
```
Comment (hugodelahousse) [inline package.json]: Fix indentation to 2 spaces please

File: src/config.ts [hunk 9]
```diff
@@ -1,9 +1,8 @@
 // Load config from `.env` file
 require('dotenv').config();
 
-function getEnv(variable: string, defaultValue = undefined): string {
-  let value = process.env[variable];
-  if (value === undefined) { value = defaultValue; }
+function getEnv(variable: string): string {
```
Comment (hugodelahousse) [inline src/config.ts]: Don't remove the default value

File: src/controller/StoreController.ts [hunk 10]
```diff
@@ -8,13 +8,16 @@ import {
   Post,
   Patch,
   HttpCode,
+  UseBefore,
   QueryParam,
   OnUndefined,
   BadRequestError,
 } from 'routing-controllers';
-import { Store } from '@entities/Store';
+import { Store } from '../entities/Store';
```
Comment (hugodelahousse) [inline src/controller/StoreController.ts]: Use @entities import

File: src/controller/UserController.ts [hunk 11]
```diff
@@ -1,17 +1,9 @@
...
-  Body,
-  Delete,
-  Get, HttpCode,
-  JsonController,
-  OnUndefined,
-  Param,
-  Patch,
-  Post,
-} from 'routing-controllers';
+import { Body, Delete, Get, JsonController, OnUndefined, Param, Post, UseBefore, BadRequestError, Patch, HttpCode } from 'routing-controllers';
```
Comment (hugodelahousse) [inline src/controller/UserController.ts]: It was better the way it was before

File: src/tests/controllers/Promotion.test.ts [hunk 12]
```diff
@@ -1,19 +1,19 @@
 import * as chai from 'chai';
 import chaiHttp = require('chai-http');
-import { app, promotions } from '../setup';
+import { app, jwtToken, promotions } from '../setup';
```
Comment (MrColorado) [inline src/tests/controllers/Promotion.test.ts]: You are missing the trailing comma that hugo love

File: src/tests/controllers/ShoppingList.test.ts [hunk 13]
```diff
@@ -1,19 +1,19 @@
 import * as chai from 'chai';
 import chaiHttp = require('chai-http');
-import { app, lists, users } from '../setup';
+import { app, jwtToken, lists, users } from '../setup';
```
Comment (MrColorado) [inline src/tests/controllers/ShoppingList.test.ts]: You are missing the trailing comma that hugo love

File: src/tests/controllers/Store.test.ts [hunk 14]
```diff
@@ -1,19 +1,21 @@
 import * as chai from 'chai';
 import chaiHttp = require('chai-http');
-import { app, stores } from '../setup';
+import { app, jwtToken, stores } from '../setup';
```
Comment (MrColorado) [inline src/tests/controllers/Store.test.ts]: You are missing the trailing comma that hugo love

**Pull request merged**